kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/parquet/ParquetStatsReader.java

```diff
@@ -0,0 +1,238 @@
+            return Literal.ofFloat(((Number) statValue).floatValue());
+        } else if (dataType instanceof DoubleType) {
+            return Literal.ofDouble(((Number) statValue).doubleValue());
+        } else if (dataType instanceof DecimalType) {
+            LogicalTypeAnnotation logicalType = statistics.type().getLogicalTypeAnnotation();
+            checkArgument(
+                    logicalType instanceof DecimalLogicalTypeAnnotation,
+                    "DECIMAL column had invalid Parquet Logical Type: %s", logicalType);
+            int scale = ((DecimalLogicalTypeAnnotation) logicalType).getScale();
+
+            DecimalType decimalType = (DecimalType) dataType;
+
+            // Check the scale is same in both the Delta data type and the Parquet Logical Type
+            checkArgument(
+                    scale == decimalType.getScale(),
+                    "DECIMAL column has different scale compared to the input data: %s", scale);
+
+            // Decimal is stored either as int, long or binary. Decode the stats accordingly.
+            BigDecimal decimalStatValue;
+            if (statistics instanceof IntStatistics) {
+                decimalStatValue = BigDecimal.valueOf((Integer) statValue).movePointLeft(scale);
+            } else if (statistics instanceof LongStatistics) {
+                decimalStatValue = BigDecimal.valueOf((Long) statValue).movePointLeft(scale);
+            } else if (statistics instanceof BinaryStatistics) {
+                BigInteger base = new BigInteger(getBinaryStat(statistics, decodeMin));
+                decimalStatValue = new BigDecimal(base, scale);
+            } else {
+                throw new UnsupportedOperationException(
+                        "Unsupported stats type for Decimal: " + statistics.getClass());
+            }
+            return Literal.ofDecimal(decimalStatValue, decimalType.getPrecision(),
+                    decimalType.getScale());
+        } else if (dataType instanceof DateType) {
+            checkArgument(
+                    statistics instanceof IntStatistics,
+                    "Column with DATE type contained invalid statistics: %s", statistics);
+            return Literal.ofDate((Integer) statValue); // stats are stored as epoch days
+        } else if (dataType instanceof StringType) {
+            byte[] binaryStat = getBinaryStat(statistics, decodeMin);
+            return Literal.ofString(new String(binaryStat, UTF_8));
+        } else if (dataType instanceof BinaryType) {
+            return Literal.ofBinary(getBinaryStat(statistics, decodeMin));
+        }
+
+        throw new IllegalArgumentException("Unsupported stats data type: " + statValue);
+    }
+
+    private static Optional<Statistics<?>> mergeMetadataList(
+            Collection<ColumnChunkMetaData> metadataList) {
+        if (hasInvalidStatistics(metadataList)) {
+            return Optional.empty();
+        }
+
+        return metadataList.stream()
```

Comment:
how hard is it to _not_ stream and reduce? `stream` is known to not have great performance

Reply:
This code is executed once per file. If it is per row, the cost adds up.